onadata/apps/api/viewsets/data_viewset.py

```diff
@@ -271,10 +271,26 @@ def enketo(self, request, *args, **kwargs):
         return Response(data=data)
 
     def destroy(self, request, *args, **kwargs):
+        instance_ids = request.data.get('instance_ids')
         self.object = self.get_object()
 
         if isinstance(self.object, XForm):
-            raise ParseError(_(u"Data id not provided."))
+            if not instance_ids:
+                raise ParseError(_(u"Data id(s) not provided."))
+            else:
+                instance_ids = filter(
+                    lambda x: x.isdigit(),
+                    instance_ids.split(',')
+                )
+
+                instances = Instance.objects.filter(
+                    id__in=instance_ids,
+                    xform=self.object
+                )
+
+                for instance in instances:
```

Comment:
We may want to actually delete in bulk here rather than for every single submission. `set_deleted()` calls submission count every time you call it.  Perhaps:

```python
from django.utils import timezone
...
initial_count = self.object.submission_count()
instances = Instance.objects.filter(
    id__in=instance_ids,
    xform=self.object,
    deleted_at__isnull=True  # do not update this timestamp when the record have already been deleted. 
).update(deleted_at=timezone.now())
after_count = self.object.submission_count(force_update=True)  # updates the num_of_submissions for the form.

number_of_records_deleted = initial_count - after_count
...
```

This does include the notion of indicating the number of records that have been deleted, we may want to respond with a `status_code = 200` in this case since we are returning this info in the response.

Reply:
Cool, that makes sense. I'll make the change.